src/graph/optimizer/rule/OptimizeTagIndexScanByFilterRule.cpp

```diff
@@ -94,9 +110,38 @@ StatusOr<TransformResult> OptimizeTagIndexScanByFilterRule::transform(
 
   OptimizerUtils::eraseInvalidIndexItems(scan->schemaId(), &indexItems);
 
+  auto condition = filter->condition();
+  auto conditionType = condition->kind();
+  Expression* transformedExpr = condition->clone();
+
+  // Stand alone IN expr with only 1 element in the list, no need to check index
+  if (conditionType == ExprKind::kRelIn) {
+    transformedExpr = graph::ExpressionUtils::rewriteInExpr(condition);
+    DCHECK(transformedExpr->kind() == ExprKind::kRelEQ);
+  }
+
+  // case2: logical AND expr
+  if (condition->kind() == ExprKind::kLogicalAnd) {
+    for (auto& operand : static_cast<const LogicalExpression*>(condition)->operands()) {
+      if (operand->kind() == ExprKind::kRelIn) {
+        auto inExpr = static_cast<RelationalExpression*>(operand);
+        // Do not apply this rule if the IN expr has a valid index or it has only 1 element in the
+        // list
+        if (static_cast<ListExpression*>(inExpr->right())->size() > 1) {
+          return TransformResult::noTransform();
+        } else {
+          transformedExpr = graph::ExpressionUtils::rewriteInExpr(condition);
+        }
+        if (OptimizerUtils::relExprHasIndex(inExpr, indexItems)) {
+          return TransformResult::noTransform();
+        }
+      }
+    }
+  }
+
```

Comment:
IN this case, we can't determine that this expression is introduced by IN, for example:
where in (....) ---> correct
where a AND (b OR c) AND ... ---> incorrect

Reply:
This is true.
This rule is to match AND expr containing IN expr as operand. i.g. `a AND (b IN [1,2,3])`

Actually, I don't really want to expand all AND expr to OR expr, since we don't know whether the property in the OR expression has a valid index or not unless we iterate all its operands. However, for IN expr `a IN [b,c,d,....]`, we only need to check the index of the left operand `a`.